org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/selectionRange/LSPSelectionRangeAbstractHandler.java

```diff
@@ -0,0 +1,244 @@
+			if (handler == null) {
+				handler = new SelectionRangeHandler(styledText);
+			}
+			return handler;
+		}
+
+		public SelectionRangeHandler(StyledText styledText) {
+			this.styledText = styledText;
+			styledText.setData(KEY, this);
+			styledText.addCaretListener(new CaretListener() {
+
+				@Override
+				public void caretMoved(CaretEvent arg0) {
+					if (!updating) {
+						root = null;
+					}
+				}
+			});
+		}
+
+		public SelectionRange getSelectionRange(Direction direction) {
+			if (direction == Direction.UP) {
+				if (previous != null) {
+					previous = previous.getParent();
+					return previous;
+				}
+			} else {
+				if (previous != null) {
+					SelectionRange selectionRange = root;
+					while (selectionRange != null) {
+						SelectionRange parent = selectionRange.getParent();
+						if (previous.equals(parent)) {
+							previous = selectionRange;
+							return previous;
+						}
+						selectionRange = parent;
+					}
+				}
+			}
+			return null;
+		}
+
+		public boolean isDirty() {
+			return root == null;
+		}
+
+		public void updateSelection(ISelectionProvider provider, IDocument document, Direction direction) {
+			if (styledText.isDisposed()) {
+				return;
+			}
+			SelectionRange selectionRange = getSelectionRange(direction);
+			if (selectionRange != null) {
+				ISelection selection = LSPEclipseUtils.toSelection(selectionRange.getRange(), document);
+				styledText.getDisplay().execute(() -> {
+					try {
+						updating = true;
+						provider.setSelection(selection);
+					} finally {
+						updating = false;
+					}
+				});
+			}
+		}
+
+	}
+
+	@Override
+	public Object execute(ExecutionEvent event) throws ExecutionException {
+		IEditorPart part = HandlerUtil.getActiveEditor(event);
+		if (part instanceof ITextEditor textEditor) {
+			final ISelectionProvider provider = textEditor.getSelectionProvider();
+			if (provider == null) {
+				return null;
+			}
+			ISelection sel = provider.getSelection();
+			ITextViewer viewer = textEditor.getAdapter(ITextViewer.class);
+			if (viewer == null) {
+				return null;
+			}
+			StyledText styledText = viewer.getTextWidget();
+			if (styledText == null) {
+				return null;
+			}
+
+			if (sel instanceof ITextSelection textSelection && !textSelection.isEmpty()) {
+				IDocument document = LSPEclipseUtils.getDocument(textEditor);
+				if (document != null) {
+					LanguageServerDocumentExecutor executor = LanguageServers.forDocument(document)
+							.withCapability(ServerCapabilities::getSelectionRangeProvider);
+					if (executor.anyMatching()) {
+						// It exists a language server which supports 'textDocument/SelectionRange' LSP
+						// operation.
+						SelectionRangeHandler.Direction direction = getDirection();
+						// Get the SelectionRangeHandler instance of the editor.
+						SelectionRangeHandler handler = SelectionRangeHandler.getSelectionRangeHandler(styledText);
+						if (handler.isDirty()) {
+							// The user:

```

Comment:
should this comment not rather go to the javadoc of the isDirty method?

Reply:
fixed